packages/mui-base/src/MenuUnstyled/MenuButtonUnstyled.tsx

```diff
@@ -0,0 +1,111 @@
+import * as React from 'react';
+import {
+  unstable_useForkRef as useForkRef,
+  unstable_useControlled as useControlled,
+} from '@mui/utils';
+import { useSlotProps } from '../utils';
+
+export interface MenuButtonProps {
+  className?: string;
+  label?: string;
+  open?: boolean;
+  onOpenChange?: (
+    event: React.MouseEvent<HTMLButtonElement> | React.KeyboardEvent<HTMLButtonElement>,
+    open: boolean,
+  ) => void;
+  defaultOpen?: boolean;
+  children?: React.ReactNode;
+  slots?: {
+    root?: React.ElementType;
+  };
+  slotProps?: {
+    root?: React.HTMLAttributes<HTMLButtonElement>;
+  };
+}
+
+interface MenuTriggerContextValue {
+  open: boolean;
+  buttonRef: React.RefObject<HTMLButtonElement>;
+}
+
+export const MenuTriggerContext = React.createContext<MenuTriggerContextValue | null>(null);
+
+const MenuButton = React.forwardRef(function MenuButton(
+  props: MenuButtonProps,
+  ref: React.ForwardedRef<HTMLButtonElement>,
+) {
+  const {
+    children,
+    label,
+    open: openProp,
+    defaultOpen,
+    onOpenChange,
+    slots = {},
+    slotProps = {},
+    ...other
+  } = props;
+
+  const [isOpen, setOpen] = useControlled({
+    controlled: openProp,
+    default: defaultOpen,
+    name: 'MenuButton',
+  });
+
+  const buttonRef = React.useRef<HTMLButtonElement>(null);
+  const handleRef = useForkRef(buttonRef, ref);
+
+  const contextValue: MenuTriggerContextValue = React.useMemo(
+    () => ({ open: isOpen, buttonRef }),
+    [isOpen, buttonRef],
+  );
+
+  const handleClick = React.useCallback(
+    (event: React.MouseEvent<HTMLButtonElement>) => {
+      setOpen((open) => {
+        return !open;
+      });
+
+      onOpenChange?.(event, !isOpen);
+    },
+    [isOpen, setOpen, onOpenChange],
+  );
+
+  const handleKeyDown = React.useCallback(
+    (event: React.KeyboardEvent<HTMLButtonElement>) => {
+      if (event.key === 'ArrowDown') {
+        event.preventDefault();
+        setOpen(true);
+        onOpenChange?.(event, true);
+      }
+    },
+    [setOpen, onOpenChange],
+  );
+
+  React.useEffect(() => {
+    if (!isOpen) {
+      buttonRef.current?.focus();
+    }
+  }, [isOpen]);
+
+  const Root = slots.root || 'button';
```

Comment:
Only thing I am not sure here is the name `root`. I would imagine the `MenuButton` to have slots `menu` and `button`, or if we want to be more generic `menu` and `trigger`. Considering the menu is the `children` element, I would rename the `slot.root` to `slot.trigger`.

Reply:
Right, it makes sense. With a small caveat - we usually forward extra props to the root slot. I wonder where they should go in this case. I'd say the `button` slot, but this will have to be well documented.

Reply:
hmm, isn't it better to keep only the `root` slot for the `MenuButtonUnstyled` for consistency?

From my understanding, MUI Base provides `slots` prop for inserting components (usually with styles). The `root` refers to the topmost DOM of the component.

Changing it to `trigger` will make the concept inconsistent and developers have to remember one more thing.